Change the ObjectToTypeName value converter to give prettier C# names.

dfkeenan opened the pull request:
# PR Details

This PR changes the `ObjectToTypeName` value converter to convert `object` `Type` to a prettier C# like style. 

## Description

This change makes the `ObjectToTypeName` value converter produce more meaningful names. The `Type` name returned has the following properties:

- Generic type parameters are included using C# like syntax. i.e. `<Type1, Type2>` instead of `'2`.
- Primitive types use the keyword syntax. i.e. `int` instead of `Int32`.
- `Nullable` types display using C# nullable syntax. i.e. `DateTime?` instead of `Nullable<DateTime>`.
- `ValueTuple` types display using C# tuple syntax i.e. `(int, string)` instead of `ValueTuple<int, string>`.
- Type names are not `namespace` or `class` name qualified. i.e. `DateTime` instead of `System.DateTime`. 

## Related Issue

None

## Motivation and Context

Currently when displaying a types name in the property grid it does not show meaningful information for generic types. For example a property of type `MyCustomClass<int?>` gets displayed as `MyCustomClass'1`. This is not very helpful. The goal of this change is to make it return more meaningfull names such as `MyCustomClass<int?>`.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

dfkeenan commented:
I currently have a few outstanding issues. 

- For some reason unit tests don't run. I get messages like `Test project Xenko.Core.Presentation.Tests does not reference any .NET NuGet adapter. Test discovery or execution might not work for this project.`. So I have not run existing tests or created new ones.
- I am not sure if the converter should keep a cached version of type names, For example keep a static `ConcurrentDictionary<Type, string>` and a static `CSharpCodeProvider`.

dfkeenan commented:
@xen2 , Do we need to install the `xunit.runner.visualstudio` NuGet package in the test projects now? See https://xunit.github.io/docs/getting-started/netfx/visual-studio#run-tests-visualstudio.

xen2 commented:
@dfkeenan 
This is supposedly already added, but might be missing from some test projects?
```
samples/Tests/Xenko.Samples.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/assets/Xenko.Core.Assets.Quantum.Tests/Xenko.Core.Assets.Quantum.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/assets/Xenko.Core.Assets.Tests/Xenko.Core.Assets.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/buildengine/Xenko.Core.BuildEngine.Tests/Xenko.Core.BuildEngine.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/core/Xenko.Core.Design.Tests/Xenko.Core.Design.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/core/Xenko.Core.Mathematics.Tests/Xenko.Core.Mathematics.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/core/Xenko.Core.Tests/Xenko.Core.Tests.Android.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/core/Xenko.Core.Tests/Xenko.Core.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/core/Xenko.Core.Tests/Xenko.Core.Tests.iOS.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/core/Xenko.Core.Yaml.Tests/Xenko.Core.Yaml.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/editor/Xenko.Core.Assets.Editor.Tests/Xenko.Core.Assets.Editor.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/editor/Xenko.GameStudio.Tests/Xenko.GameStudio.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/engine/Xenko.Assets.Tests/Xenko.Assets.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/engine/Xenko.Graphics.Regression/Xenko.Graphics.Regression.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/engine/Xenko.Shaders.Tests/Xenko.Shaders.Tests.Windows.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/presentation/Xenko.Core.Presentation.Quantum.Tests/Xenko.Core.Presentation.Quantum.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/presentation/Xenko.Core.Quantum.Tests/Xenko.Core.Quantum.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/tools/Xenko.Code.Tests/Xenko.Code.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
sources/tools/Xenko.Core.ProjectTemplating.Tests/Xenko.Core.ProjectTemplating.Tests.csproj:    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
```

xen2 commented on sources/presentation/Xenko.Core.Presentation/ValueConverters/ObjectToTypeName.cs:
```diff
@@ -19,11 +22,22 @@ public class ObjectToTypeName : OneWayValueConverter<ObjectToTypeName>
+            var fullTypeName = codeProvider.GetTypeOutput(new CodeTypeReference(valueType));
+
+            //Change System.Nullable to "Type?" format
+            var typeName = Regex.Replace(fullTypeName, @"System\.Nullable<([^>]*)>", "$1?");
```
Will this work with `Nullable<Dictionary<int, int>>` ?

dfkeenan replied:
Not sure, that's why we need unit tests. ;)

I will add tests for nested Generic types.

dfkeenan replied:
I might have to get crazy with regex. https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#balancing-group-definitions

I might do similar for `System.ValueTuple`. Not sure how well such types work in the Property grid though.

xen2 replied:
I wouldn't recommend playing with string regex for that.
We have a proper source `Type` so use this as source rather than reprocessing a string and doing it in two steps (esp. if the second step could be quite error-prone, we almost need a parser to properly understand generic depth, etc.).

Did you check if there wasn't a library that could do that automatically for us?
FYI we already link Roslyn (Microsoft.CodeAnalysis).

dfkeenan replied:
My initial searching got me to `CSharpCodeProvider.GetTypeOutput` method. I'll do a bit more research.

dfkeenan replied:
Re-wrote ObjectToTypeConverter to use a custom method that handles nullables and tuples better

dfkeenan commented:
Looks like it is missing in the one test project I was looking at `Xenko.Core.Presentation.Tests`.  I will add it to that.

dfkeenan commented:
I have added unit tests that cover a fair few cases.  I have checked the `All new and existing tests passed.` criteria. But I only ran the tests in the test project `Xenko.Core.Presentation.Tests` for the assembly I changed.

xen2 commented on sources/presentation/Xenko.Core.Presentation/ValueConverters/ObjectToTypeName.cs:
```diff
@@ -17,13 +22,87 @@ public class ObjectToTypeName : OneWayValueConverter<ObjectToTypeName>
         /// The string representation of the type of a null object
         /// </summary>
-        public const string NullObjectType = "(None)";
+        public const string NullObjectType = "(None)";        
```
Please remove end of line spaces

xen2 commented on sources/presentation/Xenko.Core.Presentation/ValueConverters/ObjectToTypeName.cs:
```diff
@@ -17,13 +22,87 @@ public class ObjectToTypeName : OneWayValueConverter<ObjectToTypeName>
+            if (value == null) return NullObjectType;
+
+            return value.GetType().ToSimpleCSharpName();
+            
```
Please remove empty line before brace

dfkeenan replied:
Did you mean just delete line 34?

🤔 I could also change it to be more like the original.

```cs
return value?.GetType().ToSimpleCSharpName() ?? NullObjectType;
```

xen2 replied:
Ha yes, even better.

dfkeenan replied:
I have done this.

xen2 commented on sources/presentation/Xenko.Core.Presentation/ValueConverters/ObjectToTypeName.cs:
```diff
@@ -17,13 +22,87 @@ public class ObjectToTypeName : OneWayValueConverter<ObjectToTypeName>
+    /// <summary>
+    /// Helper class to get formatted names from <see cref="Type"/>.
+    /// </summary>
+    public static class TypeNameHelper
```
Either it should be a private nested type of ObjectToTypeName class, or be in its own separate file.

dfkeenan replied:
I mostly made it a separate public class for unit test purposes. If we keep it public I guess I'll need to find a good home for it. It is a fairly generic helper class so could possibly be used elsewhere. 🤔

dfkeenan replied:
Maybe I'll put it with the other extensions in the `Xenko.Core.Presentation.Extensions` namespace.

xen2 replied:
You can make it internal, and add `InternalsVisibleTo`  attribute in `AssemblyInfo.cs` for unit test purpose.

dfkeenan replied:
OK, I'll move it to a separate file in `Xenko.Core.Presentation.Extensions`, and make it internal etc.

dfkeenan replied:
I have moved `TypeNameHelper` to it's own file in `Xenko.Core.Presentation.Extensions` and renamed it to `TypeExtensions` to make it more inline with the rest of the classes in that namespace.  I also moved it's tests to it's own unit test class.

dfkeenan commented:
@xen2 , Would you prefer a commit per review issue or is 1 "Resolve review issues" commit OK?

xen2 commented:
> @xen2 , Would you prefer a commit per review issue or is 1 "Resolve review issues" commit OK?

I will probably do a squash before merge so you don't need to bother.

dfkeenan commented:
> I will probably do a squash before merge so you don't need to bother.

OK!

dfkeenan commented:
I think I have addressed all the review issues. I wasn't sure who is supposed to "Resolve conversation" so I haven't for the time being.

xen2 commented on sources/presentation/Xenko.Core.Presentation.Tests/Xenko.Core.Presentation.Tests.csproj:
```diff
@@ -12,6 +12,10 @@
     <PackageReference Include="xunit" Version="2.3.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
```
It seems all the other references to `xunit.runner.visualstudio` don't have IncludeAssets/PrivateAssets.
Let's simplify those four lines to `<PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />`

dfkeenan replied:
OK, will do.

xen2 commented on sources/presentation/Xenko.Core.Presentation/Extensions/TypeExtensions.cs:
```diff
@@ -0,0 +1,79 @@
+                return fullTypeName.Substring(simpleNameStart);
+            }
+
+            if(Nullable.GetUnderlyingType(type) is Type nullableType)
```
space between if and parenthesis

dfkeenan replied:
I thought I fixed all my formatting issues. 👀 
Will do..

dfkeenan commented:
OK, I have done those 2 things.

xen2 commented:
Merged, thanks!
